### ISSUE TITLE
Is this a bug that `Enumerable.Repeat(T, int)` produce an `IEnumerable<T?>` as return type?

### ISSUE THREAD

**Issue by qaqz111** [dotnet/roslyn#49977]
**Version Used**: 16.8.3

**Steps to Reproduce**:

1. Code:
```C#
        public static T[] ExpandTo<T>(this T[] array, int length, T val = default)
            => 0 >= (length -= array.Length)
                ? array
                : array.Concat(Enumerable.Repeat(val, length)).ToArray();
```
2. Try to compile this code, the compiler give a warning CS8620: Argument of type 'IEnumerable<T?>' cannot be used for parameter 'second' of type 'IEnumerable<T>' in 'IEnumerable<T> Enumerable.Concat<T>(IEnumerable<T> first, IEnumerable<T> second)' due to differences in the nullability of reference types.

The call of `Enumerable.Repeat(val, length)` was marked by wave lines.
This warning appeared for quite a while and does not get fixed for now.
I searched github issues in Roslyn, csharplang, corefx projects, is this the same issue as #37468?

Why the returned type of `Enumerable.Repeat` is `IEnumerable<T?>` while there's no constrait of `T : class`?

**Comment by huoyaoyuan** [comment]
This should be caused by `T val = default` I think?

**Comment by RikkiGibson** [comment]
In 16.8 we're giving `val` a maybe-default initial state because of the `default` parameter value. This causes us to infer a `T?` type argument to `Enumerable.Repeat`. In 16.9 we don't modify parameter initial state based on default parameter values--we just warn on any bad default values, so you'd only see a warning on the `default` literal itself.

`default(T)` is only assignable to a value of type `T?` when `T` is unconstrained. This method as written probably needs to return type `T?[]` as written in order to be null-safe. However, even if you fix the method to use `T? val` and return `T?[]`, there's actually still a bug with type inference to `Concat` (seemingly?). Illustrated in this [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABAJgEZGmAGCptgFgDcnHnwB0AGQCWuAI7CGnAMy8WFAMIUA3owp7eKvkoA8fLgH4AfBQCybABRHTbLtbhskvF+arkwASm1dfRCmAHYqNjF1AHtcAGMIGHsAUQIwRFo4MQAlOAAHOCT7OD9PNn9/BRCAX2C9eoMvEwAVK1sWYxbLRzZWtw8KNt84AKCGENCI92i4xOS0/AyELNyCouTS0fLK6v06if1GpkM+rvabft7+yM9hrbGAdwALREoWihBVccn9cMjZgliotlqs8oVig8dlVGgcQsdTq0Lvwuj0nN1bkMfA9Ai83kNPqofDpDr9/jNYkCFulMmRsuCNiUyhQKjDSRQDjUgA==). Note that only M2 (unconstrained T) gets a warning here, and not the `class` or `class?` constrained versions. That makes me wonder if it's a type substitution bug similar to #41368.

As it stands I recommend you rewrite the method [as follows](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChGABABguYEYA6AGQEtcARwDcjFgGYOnAGwcATBQDCFAN6MKmjlK5yAKgH4A2gF0KAUQAeABwi4UegPYAePQD4AFDAAW/KBT1TCggEBAgATyQKQRgKclwAcx8owwoANwhCCgBeChQ4ADMIIhgASg0tSuy3CnY3XI94pO8KAFpckLDwvmpm8oZKwa0DYNCIiqHKkFGu7iVHXABjCBgPcwIwRFo4bgAlOGs4FY8Mwiimn1KAQlLuJwBBMfCPUrEGAF9GIA):
```cs
    public static T?[] ExpandTo<T>(this T[] array, int length, T? val = default)
        => 0 >= (length -= array.Length)
            ? array
            : array.Concat(Enumerable.Repeat(val, length)!).ToArray();
```

**Comment by jaredpar** [comment]
@RikkiGibson do we feel confident the remaining issue is a dupe of #41368 (enough to close). Or is more investigation needed here?

**Comment by RikkiGibson** [comment]
The bad behavior is still present, so I don't consider this a dupe, more that there is a change needed to fix this which may be analogous to the change that fixed #41368.

**Comment by qaqz111** [comment]
Will `defualt!` affect the compiler's type inference?
I tried this:
```C#
#nullable enable

using System.Linq;

public static class C {
    public static void foo1(){
        var k = new object[1].ExpandTo(9);
        object k1 = k[0]; // CS8600
    }

    public static void foo2(){
        var k = new object[1].ExpandTo<object>(9); //<object> is marked as redundant, but without this
                                                                                // the compiler report CS8600 on next line as in foo1()
        object k1 = k[0]; //OK
    }

    public static T[] ExpandTo<T>(this T[] array, int length, T val = default!) //defualt! suppressed CS8620, but looks like
                                                                                                                       // the compiler still infer an T?[] output?
        => 0 >= (length -= array.Length)
            ? array
            : array.Concat(Enumerable.Repeat(val, length)).ToArray();
}
```

**Comment by RikkiGibson** [comment]
IIRC, it doesn't affect the initial state of 'val' due to limitations around round-tripping from BoundExpression to ConstantValue back to BoundExpression. The bug is obviated by changes made in 16.9.

**Comment by qaqz111** [comment]
Understand, thanks.

**Comment by RikkiGibson** [comment]
Thanks for filing this issue. I believe the bug in the issue description has been fixed. Please feel free to follow up here if necessary.
